### ISSUE TITLE
Constructor injection can fail on circular imports

### ISSUE THREAD

**Issue by justinmchase** [typestack/typedi#96]
If two files essentially have a circular import then typescript will generate code that injects undefined metadata into paramtypes. When this happens typedi will fail to resolve the parameter type silently and can cause hard to debug errors.

For example this snippet of code between two classes in two different files with a circular import reference:
```js
ExampleB = __decorate([
    typedi_1.Service(),
    __metadata("design:paramtypes", [ExampleA_1.ExampleA])
], ExampleB);
exports.ExampleB = ExampleB;
``` 

But because ExampleA and ExampleB have circular imports, `ExampleA_1.ExampleA` is undefined at this time and no error is thrown.

Later when instantiated the service paramters are not resolved and injected properly:

```ts
@Service()
export class ExampleB {
  constructor(exa: ExampleA) {
    console.log(exa) // undefined
  }
}
```

Obviously typedi can't solve this issue magically but it should throw an error that more accurately explains what is happening and how to solve it because as it is, with a complex app, it can be really hard to understand why the injection is failing.

**Comment by vbraun** [comment]
+1 to better error messages

However, instead of relying to typedi to diagnose circular imports I prefer to push it up to the build step, for example with the webpack circular-dependency-plugin. That catches all circular imports, not just the ones involving typedi.

**Comment by justinmchase** [comment]
@vbraun This is kind of what we ended up doing also. Unfortunately we have existing circular dependencies which are unfixable but I made a CI step that checks the constructors of @Service's and detects circular references and reports errors. It works well.

Sadly, my final conclusion about this is that you should really never use constructor injection with typedi. It's essentially relying on the typescript decorator metadata, which is frankly broken and flawed and unusable in its current form.

The fact is that circular dependencies are supported by nodejs and they do resolve properly but the decorator metadata system improperly references the dependencies syncronously as the module is loaded rather than to defer till after the modules are loaded. This is out of the scope for typedi to solve probably and so the constructor injection can't be relied on to work properly. The only technique that works consistently is something like this:

```
import { B } from "./b"

@Service()
export class A {
  private b = Container.get(B)
}
```

That will always work because it will reference B at the time of construction rather than at module load time. Its less _pretty_ but I think that typedi should consider dropping support for constructor injection until / unless the decorator issues are resolved by the community.

**Comment by vbraun** [comment]
I see your point, but I'd argue that dependency injection without a clear understanding of who depends on what is always going to cause pain in the long run. Eventually you'll create actual dependency cycles of the form
~~~~
@Service()
export class A {
  private b = Container.get(B)
}

@Service()
export class B {
  private a = Container.get(A)
}
~~~~
which just can't work but will manifest in rather spurious errors. Whether or not you use constructor injection doesn't matter. Since you must have an order relation on the dependencies its never **that** difficult to actually break import cycles. Typescript elides imports that are types only, so you only need to break cycles for imports that end up in value positions.

**Comment by NoNameProvided** [comment]
We cannot prevent users from creating circular dependencies but after #87 is landed we will throw errors whenever a service cannot be resolved. 

Closing in favor of #87.
